README.txt
lizard-sticky
==========================================

Introduction


Requirements
------------

Make sure you have a geo enabled database. A good example is
postGIS. This app will NOT work on sqlite.


Usage
-----

- Add lizard_sticky to your buildout.cfg.

- Add 'lizard_sticky' and 'lizard_map' to the INSTALLED_APPS in your
  settings.

- Include lizard-sticky in your urls.py.
    (r'^sticky/', include('lizard_sticky.urls')),

- Refer to the sticky browser from somewhere:

{% url lizard_sticky.sticky_browser %}

or

reverse('lizard_sticky.sticky_browser')

- Run syncdb to make appropriate database tables.


Development installation
------------------------

The first time, you'll have to run the "bootstrap" script to set up setuptools
and buildout::

    $> python bootstrap.py

And then run buildout to set everything up::

    $> bin/buildout

(On windows it is called ``bin\buildout.exe``).

You'll have to re-run buildout when you or someone else made a change in
``setup.py`` or ``buildout.cfg``.

The current package is installed as a "development package", so
changes in .py files are automatically available (just like with ``python
setup.py develop``).

If you want to use trunk checkouts of other packages (instead of released
versions), add them as an "svn external" in the ``local_checkouts/`` directory
and add them to the ``develop =`` list in buildout.cfg.

Tests can always be run with ``bin/test`` or ``bin\test.exe``.


CHANGES.rst
Changelog of lizard-sticky
===================================================


2.2 (unreleased)
----------------

- Nothing changed yet.


2.1 (2013-03-29)
----------------

- Beautified the admin interface: showing more columns in the list interface
  and allowing filtering on tag.


2.0 (2013-03-28)
----------------

- Moved to github: https://github.com/lizardsystem/lizard-sticky.

- **Upgrade warning**: south migrations added. For existing sites, you'll need
  a ``bin/django migrate --fake lizard_sticky 0001``.


1.13 (2012-12-19)
-----------------

- Fixed urls.py, so it won't recusively include other lizard-* URLs when
  running as part of a site.


1.12 (2012-12-17)
-----------------

- Minor tree styling.


1.11 (2012-11-27)
-----------------

- Convert to Lizard 4.0.

- Add support for the legend.

- Properly set dependency versions.


1.10 (2011-12-20)
-----------------

- Fixed add to collage button in popup.

- Made saving work again.

- Require latest lizard-map (3.8).


1.9 (2011-11-11)
----------------

- Modernized buildout.cfg somewhat

- Changed views to classes to work with latest lizard-map. Updated
  templates to reflect this.

- Changed setup.py to require at least the currently latest
  lizard-map (3.4.2) and lizard-ui (3.7).

- Added dependency of map >= 1.60.


1.8 (2011-04-21)
----------------

- Removed unnecessary workspace_manager and date_range_form stuff. It
  is also incompatible with map >= 1.71.


1.7.2 (2011-04-14)
------------------

- Removed popup header (looks better).


1.7.1 (2011-03-01)
------------------

- Removed print statement.


1.7 (2011-03-01)
----------------

- Fixed sticky coordinate storage.


1.6 (2011-02-28)
----------------

- Changed storage for x, y coordinates: Coordinates are
  now saved in original projection (i.e. google, rd). When drawing
  layer, tell mapnik the projection setting in MAP_SETTINGS. This also
  fixes a problem when the client is not in 900913
  projection. Requires lizard-map 1.55 or higher.


1.5 (2011-02-15)
----------------

- Moved sticky.png to app_icons subdir.


1.4 (2011-02-03)
----------------

- Fixed breadcrumbs bug.


1.3 (2011-02-01)
----------------

- Minor layout changes.

- Added option crumbs_prepend (see lizard_ui).


1.2 (2010-12-10)
----------------

- Renamed "Geeltje" to "Melding".


1.1 (2010-11-08)
----------------

- Updated readme.

- Added sticky icon.

- Bugfix draw multiple stickies close to each other to compensate #402.


1.0 (2010-09-22)
----------------

- Bugfix posting stickies.

- Clean up some code.


0.8 (2010-09-22)
----------------

- Bugfix: add preventDefault to js. This will prevent the js to reload
  the page after posting a new sticky.


0.7 (2010-09-22)
----------------

- Bugfix: "show all stickies" was broken.

- Bugfix: better compensation for mapnik bug #402.


0.6 (2010-09-22)
----------------

- Bugfix: prevent javascript to reload page (by adding "return
  false").

- Bugfix: layers that filters tags now work.

- Bugfix: do not add "empty" tags.


0.5 (2010-09-03)
----------------

- Removed mapnik custom database query stuff, replaced by standard
  django requests and building points layer manually.


0.4 (2010-09-01)
----------------

- Fixed stickies by searching just by ID instead of by elementtype#id
  combinations.  The bug was that one of the elements changed type recently.


0.3 (2010-08-30)
----------------

- Refactor add-sticky to new popup style.


0.2 (2010-08-18)
----------------

- Adjusted to lizard-map's adapter changes and snippet group functionality.


0.1 (2010-07-15)
----------------

- Tags can be used as filters on sticky layer.
- Navigate through stickies and put new stickies using the map.
- Initial library skeleton created by nensskel.  [Jack]


lizard_sticky/layers.py
"""
Adapter for lizard-sticky
"""
import os
import logging
import mapnik

from django.conf import settings
from django.contrib.gis.geos import Point
from django.contrib.gis.measure import D
from django.shortcuts import get_object_or_404
from django.template.loader import render_to_string
from django.utils import simplejson as json

from lizard_map import workspace
from lizard_map.coordinates import WGS84
from lizard_map.coordinates import wgs84_to_google
from lizard_map.models import ICON_ORIGINALS
from lizard_map.symbol_manager import SymbolManager
from lizard_sticky.models import Sticky
from lizard_sticky.models import Tag

ICON_STYLE = {'icon': 'sticky.png',
              'mask': ('sticky_mask.png', ),
              'color': (1, 1, 0, 0)}

logger = logging.getLogger('lizard_sticky.layers')


class WorkspaceItemAdapterSticky(workspace.WorkspaceItemAdapter):

    def __init__(self, *args, **kwargs):
        """
        tags: list or queryset of tags

        If no tags are selected, all stickies are selected!
        """
        super(WorkspaceItemAdapterSticky, self).__init__(*args, **kwargs)
        self.tags = []
        self.tag_objects = []
        if 'tags' in self.layer_arguments:
            self.tags = self.layer_arguments['tags']
            self.tag_objects = [Tag.objects.get(slug=tag) for tag in self.tags]

    def style(self):
        """
        Make mapnik point style
        """
        symbol_manager = SymbolManager(
            ICON_ORIGINALS,
            os.path.join(settings.MEDIA_ROOT, 'generated_icons'))
        output_filename = symbol_manager.get_symbol_transformed(
            ICON_STYLE['icon'], **ICON_STYLE)
        output_filename_abs = os.path.join(
            settings.MEDIA_ROOT, 'generated_icons', output_filename)
        # use filename in mapnik pointsymbolizer
        point_looks = mapnik.PointSymbolizer(
            output_filename_abs, 'png', 16, 16)
        point_looks.allow_overlap = True
        layout_rule = mapnik.Rule()
        layout_rule.symbols.append(point_looks)
        point_style = mapnik.Style()
        point_style.rules.append(layout_rule)

        return point_style

    def layer(self, layer_ids=None, request=None):
        """Return a layer with all stickies or stickies with selected
        tags
        """

        # Use these coordinates to put points 'around' actual
        # coordinates, to compensate for bug #402 in mapnik.
        around = [(0.00001, 0),
                  (-0.00001, 0),
                  (0, 0.00001),
                  (0, -0.00001)]

        layers = []
        styles = {}
        layer = mapnik.Layer("Stickies", WGS84)

        layer.datasource = mapnik.PointDatasource()
        if self.tag_objects:
            stickies = Sticky.objects.filter(tags__in=self.tag_objects)
        else:
            stickies = Sticky.objects.all()

        for sticky in stickies:
            layer.datasource.add_point(
                sticky.geom.x, sticky.geom.y, 'Name', str(sticky.title))
            for offset_x, offset_y in around:
                layer.datasource.add_point(
                    sticky.geom.x + offset_x,
                    sticky.geom.y + offset_y,
                    'Name',
                    str(sticky.title))

        # generate "unique" point style name and append to layer
        style_name = "Stickies"
        styles[style_name] = self.style()
        layer.styles.append(style_name)

        layers = [layer, ]
        return layers, styles

    def values(self, identifier, start_date, end_date):
        """Return values in list of dictionaries (datetime, value, unit)
        """

        stickies = Sticky.objects.filter(datetime__gte=start_date,
                                         datetime__lte=end_date)
        return [{'datetime': sticky.datetime,
                 'value': sticky.description,
                 'unit': ''} for sticky in stickies]

    def search(self, google_x, google_y, radius=None):
        """
        returns a list of dicts with keys distance, name, shortname,
        google_coords, workspace_item, identifier
        """
        #from lizard_map.coordinates import google_to_rd
        #x, y = google_to_rd(google_x, google_y)
        #pnt = Point(x, y, srid=28992)  # 900913
        pnt = Point(google_x, google_y, srid=900913)  # 900913
        #print pnt, radius
        stickies = Sticky.objects.filter(
            geom__distance_lte=(pnt, D(m=radius * 0.5)))

        result = [{'distance': 0.0,
                   'name': '%s (%s)' % (sticky.title, sticky.reporter),
                   'shortname': str(sticky.title),
                   'object': sticky,
                   #'google_coords': wgs84_to_google(sticky.geom.x,
                   #                                 sticky.geom.y),
                   'workspace_item': self.workspace_item,
                   'identifier': {'sticky_id': sticky.id},
                   } for sticky in stickies]
        return result

    def location(self, sticky_id, layout=None):
        """
        returns location dict.

        requires identifier_json
        """
        sticky = get_object_or_404(Sticky, pk=sticky_id)
        identifier = {'sticky_id': sticky.id}
        return {
            'name': '%s (%s)' % (sticky.title, sticky.reporter),
            'shortname': str(sticky.title),
            'workspace_item': self.workspace_item,
            'identifier': identifier,
            'google_coords': wgs84_to_google(sticky.geom.x, sticky.geom.y),
            'object': sticky,
            }

    def symbol_url(self, identifier=None, start_date=None, end_date=None):
        return super(WorkspaceItemAdapterSticky, self).symbol_url(
            identifier=identifier,
            start_date=start_date,
            end_date=end_date,
            icon_style=ICON_STYLE)

    def html(self, identifiers=None, layout_options=None):
        """
        Renders stickies
        Note: removed snipper_group kwargs because it is not used anymore.
        Therefore sticky now requires lizard_map >= 3.7, the current
        version.
        """

        display_group = [
            self.location(**identifier) for identifier in identifiers]
        add_snippet = False
        if layout_options and 'add_snippet' in layout_options:
            add_snippet = layout_options['add_snippet']
        return render_to_string(
            'lizard_sticky/popup_sticky.html',
            {'display_group': display_group,
             'add_snippet': add_snippet,
             'symbol_url': self.symbol_url(),
             'adapter_class': self.adapter_class,
             'adapter_layer_json': json.dumps(self.layer_arguments)})

    def legend_image_urls(self):
        """
        returns symbol

        TODO: the identifier is always None, so individual symbols
        cannot be retrieved.
        """
        symbol_manager = SymbolManager(
            ICON_ORIGINALS,
            os.path.join(settings.MEDIA_ROOT, 'generated_icons'))
        output_filename = symbol_manager.get_symbol_transformed(
            ICON_STYLE['icon'], **ICON_STYLE)
        icon = '%sgenerated_icons/%s' % (settings.MEDIA_URL, output_filename)
        return [icon]
